### ISSUE TITLE
ASR model to TensorRT

### ISSUE THREAD

**Issue by fushh** [NVIDIA/NeMo#2696]
Recently, I want to convert my NeMo ASR model to TensorRT. My model was trained with NeMo v0.10.2 and when I tried to quantify it to int8, it came up with an error:

> [TensorRT] WARNING: Int8 calibration in TensorRT won't be used on networks containing boolean operators or loops. Please use quantization-aware training to generate a network with Quantize/Dequantize nodes.

Then, I transform my model into NeMo v1.2.0 following [the code here](https://github.com/NVIDIA/NeMo/discussions/1680#discussioncomment-321962). It works well without any quantification. But there are still some errors when converting to int8 following [sample code](https://github.com/NVIDIA/NeMo/tree/main/examples/asr/quantization)  without any changes except model weight parameters.

> [TensorRT] ERROR: ModelImporter.cpp:720: While parsing node number 39 [QuantizeLinear -> "687"]:
[TensorRT] ERROR: ModelImporter.cpp:721: --- Begin node ---
[TensorRT] ERROR: ModelImporter.cpp:722: input: "encoder.encoder.1.mconv.1.conv.weight"
input: "684"
input: "2973"
output: "687"
name: "QuantizeLinear_39"
op_type: "QuantizeLinear"
attribute {
  name: "axis"
  i: 0
  type: INT
}
[TensorRT] ERROR: ModelImporter.cpp:723: --- End node ---
[TensorRT] ERROR: ModelImporter.cpp:726: ERROR: builtin_op_importers.cpp:1039 In function QuantDequantLinearHelper:
[6] Assertion failed: scaleAllPositive && "Scale coefficients must all be positive"
[TensorRT] ERROR: 4: [network.cpp::validate::2411] Error Code 4: Internal Error (Network must have at least one output)
[TensorRT] ERROR: 2: [builder.cpp::buildSerializedNetwork::417] Error Code 2: Internal Error (Assertion enginePtr != nullptr failed.)

And there are also many warnings like

> I0819 14:32:00.546462 140236154607360 _utils.py:72] Input is fake quantized to 8 bits in QuantConv1d with axis None!
I0819 14:32:00.546510 140236154607360 _utils.py:75] Weight is fake quantized to 8 bits in QuantConv1d with axis 0!
I0819 14:32:00.546570 140236154607360 tensor_quantizer.py:105] Creating Max calibrator
I0819 14:32:00.546631 140236154607360 tensor_quantizer.py:105] Creating Max calibrator
I0819 14:32:00.546823 140236154607360 tensor_quantizer.py:105] Creating Max calibrator

Is there anything wrong? I would appreciate it if any help is offered!

**Comment by titu1994** [comment]
@Slyne any advice?

**Comment by Slyne** [comment]
@fushh Could you give more details about the steps to reproduce this issue ? For example, what is the model and the commandline you used.  We quantize model by:
```bash
# first calibrate model
python3 speech_to_text_calibrate.py --asr_model=QuartzNet15x5Base-En --dataset=librispeech.dev.manifest.json 
# then you should get some model like QuartzNet15x5Base-En-max.nemo

# do sensitivity analysis
python3 speech_to_text_quant_infer.py --asr_model=QuartzNet15x5Base-En-max.nemo --sensitivity  --wer_target=0.07 --onnx
# then you should get an onnx model, QuartzNet15x5Base-En-max.onnx

# check if you have installed tensorrt 8.x, if the default env is tensorrt7.x, please install tensorrt8.x
# convert to tensorrt engine
python3 speech_to_text_quant_infer_trt.py --asr_model=QuartzNet15x5Base-En --dataset=librispeech.dev.manifest.json --qat --asr_onnx=QuartzNet15x5Base-En-max.onnx --qat
```

**Comment by fushh** [comment]
@Slyne My commonds are as follows:

```
python3 speech_to_text_calibrate.py --dataset=/media/dm/D/data/test_clean.json --load_dir=experiments/ckpts/
# I got QuartzNet15x5Base-En-max-4.nemo

python3 speech_to_text_quant_infer.py --asr_model=QuartzNet15x5Base-En-max-4.nemo --dataset=/media/dm/D/data/test_clean.json --sensitivity  --wer_target=0.03 --onnx
# Then I got QuartzNet15x5Base-En-max-4.onnx

python3 speech_to_text_quant_infer_trt.py --asr_model=QuartzNet15x5Base-En-max-4.nemo --dataset=/media/dm/D/data/test_clean.json --qat --asr_onnx=QuartzNet15x5Base-En-max-4.onnx
# failed with the error above ([TensorRT] ERROR: ModelImporter.cpp:722: input: "encoder.encoder.1.mconv.1.conv.weight" ... )
```
I just add the following code in [speech_to_text_calibrate.py](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/quantization/speech_to_text_calibrate.py) to use my own weights trained with NeMo v0.10.2.
```
asr_model = EncDecCTCModel.from_pretrained(model_name="QuartzNet15x5Base-En")
enc = torch.load(args.load_dir + "JasperEncoder-STEP-190000.pt")
dec = torch.load(args.load_dir + "JasperDecoderForCTC-STEP-190000.pt")
asr_model.encoder.load_state_dict(enc)
asr_model.decoder.load_state_dict(dec)
del enc
del dec
asr_model.summarize()
asr_model.save_to(args.load_dir + "model.nemo")
asr_model_cfg = EncDecCTCModel.restore_from(restore_path=args.load_dir + "model.nemo", return_config=True)
with open_dict(asr_model_cfg):
    asr_model_cfg.encoder.quantize = True
asr_model = EncDecCTCModel.restore_from(restore_path=args.load_dir + "model.nemo", override_config_path=asr_model_cfg)
```
Is there anything wrong with my code?
My local environment is

> Python v3.8.8
> CUDA v11.1.1
> cuDNN v8.2.1
> PyTorch v1.9.0
> TensorRT v8.0.1.6
> NeMo v0.10.2 / v1.2.0
> GPU GeForce GTX 1070 Ti 8G
> CPU Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz

Looking forward to your reply, thanks!

**Comment by Slyne** [comment]
@fushh 
Please try the following command:
```
trtexec --minShapes=audio_signal:1x64x512 \
              --optShapes=audio_signal:1x64x512 \
              --maxShapes=audio_signal:1x64x512 \
              --int8
              --onnx=QuartzNet15x5Base-En-max-4.nemo
```

If it doesn't work, do you mind sending me a link to your onnx model ?

**Comment by fushh** [comment]
@Slyne 
Hi, I followed your advice and tried it again, but it still failed with the same error.

[My quantification code](https://github.com/fushh/NeMo-to-TensorRT/tree/v1.0) and [onnx model](https://github.com/fushh/NeMo-to-TensorRT/releases/tag/v1.0) have uploaded to my Github. And more details are shown in `README.md`.

Maybe the error is caused by the difference of GPUs? The model was trained on RTX 3090 and the model is now converting on GTX 1070 Ti. But when I convert the model trained on titan v, it works.

Can you give me some further advice? Thank you very much for taking up your valuable time.

**Comment by fushh** [comment]
@Slyne I still have another problem. According to EncDecCTCModel.forward's interface
```
def forward(
        self, input_signal=None, input_signal_length=None, processed_signal=None, processed_signal_length=None
    ):
        """
        Forward pass of the model.

        Args:
            input_signal: Tensor that represents a batch of raw audio signals,
                of shape [B, T]. T here represents timesteps, with 1 second of audio represented as
                `self.sample_rate` number of floating point values.
            input_signal_length: Vector of length B, that contains the individual lengths of the audio
                sequences.
            processed_signal: Tensor that represents a batch of processed audio signals,
                of shape (B, D, T) that has undergone processing via some DALI preprocessor.
            processed_signal_length: Vector of length B, that contains the individual lengths of the
                processed audio sequences.

        Returns:
            A tuple of 3 elements -
            1) The log probabilities tensor of shape [B, T, D].
            2) The lengths of the acoustic sequence after propagation through the encoder, of shape [B].
            3) The greedy token predictions of the model of shape [B, T] (via argmax)
        """
```
It needs two inputs and has three outputs. But, in the [code](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/quantization/speech_to_text_quant_infer_trt.py), you feed actually one input and get one output. What's more, 
```
for test_batch in asr_model.test_dataloader():
            if can_gpu:
                test_batch = [x.cuda() for x in test_batch]
            processed_signal, processed_signal_length = asr_model.preprocessor(
                input_signal=test_batch[0], length=test_batch[1]
            )

            greedy_predictions = trt_inference(
                stream,
                trt_ctx,
                d_input,
                d_output,
                input_signal=processed_signal,
                input_signal_length=processed_signal_length,
            )
```
you seem to pass the processed_signal as input_signal. Is there anything wrong?

Looking forward to your answers to the above two questions. Thanks.

**Comment by Slyne** [comment]
@fushh Q2: Please see https://github.com/NVIDIA/NeMo/issues/2317
Q1. Sorry, I don't have 1070ti to test the model. You may pass the onnx model to tensorrt： https://github.com/NVIDIA/TensorRT/issues

**Comment by fushh** [comment]
> @fushh Q2: Please see #2317
> And the exported onnx model only contains model part not including preprocessor part. Therefore it's different from `EncDecCTCModel`.
> 
> Q1. Sorry, I don't have 1070ti to test the model. You may pass the onnx model to tensorrt： https://github.com/NVIDIA/TensorRT/issues

Thank you for your help!
